lib/engine/activity_monitor.ts

```diff
@@ -0,0 +1,113 @@
+ * inactive.
+ */
+export class ActivityMonitor extends events.EventEmitter {
+    private _running : boolean;
+    private _appdb : AppDatabase;
+    private _assistant : AssistantDispatcher;
+    private _lastUpdate : number;
+    private _idleTimeout : NodeJS.Timeout|null;
+    private _idleTimeoutMillis : number;
+    private _quiesceTimeout : NodeJS.Timeout|null;
+    private _quiesceTimeoutMillis : number;
+
+    constructor(appdb : AppDatabase, assistant : AssistantDispatcher, options : ActivityMonitorOptions) {
+        super();
+        this._running = false;
+        this._appdb = appdb;
+        this._assistant = assistant;
+        this._lastUpdate = 0;
+        this._idleTimeout = null;
+        this._idleTimeoutMillis = options.idleTimeoutMillis || DEFAULT_IDLE_TIMEOUT;
+        this._quiesceTimeout = null;
+        this._quiesceTimeoutMillis = options.quiesceTimeoutMillis || DEFAULT_QUIESCE_TIMEOUT;
+    }
+
+    start() {
+        this._appdb.on('app-added', this.updateActivity.bind(this));
+        this._appdb.on('app-removed', this.updateActivity.bind(this));
+        this._running = true;
+        console.log('Activity monitor started');
+        this.updateActivity();
+    }
+
+    updateActivity() {
+        if (!this._running)
+            return;
+        this._lastUpdate = Date.now();
+        if (this._idleTimeout !== null)
+            return;
+        if (this._quiesceTimeout !== null) {
+            clearTimeout(this._quiesceTimeout);
+            this._quiesceTimeout = null;
+        }
+        this._maybeStartActivityTimeout();
+    }
+
+    _maybeStartActivityTimeout() {
+        if (this._appdb.getAllApps().length > 0) 
+            return;
+        const idleTimeout = Math.max(this._idleTimeoutMillis - (Date.now() - this._lastUpdate), 0);
+        console.log(`Activity monitor starts idle timer ${idleTimeout} ms`);
+        this._idleTimeout = setTimeout(() => {
+            this._idleTimeout = null;
+            const msSinceLastUpdate = Date.now() - this._lastUpdate;
+            if (msSinceLastUpdate >= this._idleTimeoutMillis)
+                this._startQuiesce();
+            else
+                this._maybeStartActivityTimeout();
+        }, idleTimeout);
+    }
+
+    _startQuiesce() {
+        console.log(`Activity monitor starts quiesce timer ${this._quiesceTimeoutMillis} ms`);
+        this._quiesceTimeout = setTimeout(() => {
+            this._quiesceTimeout = null;
+            console.log('Activity monitor emits idle event');
+            this.emit("idle");
+        }, this._quiesceTimeoutMillis);
+        for (const [_, conversation] of this._assistant.getConversations())
+            conversation.sendPing();
+    }
+
+    stop() {
+        this._running = false;
+        if (this._idleTimeout !== null) 
+            clearTimeout(this._idleTimeout);
+        if (this._quiesceTimeout !== null)
+            clearTimeout(this._quiesceTimeout);
+        console.log('Activity monitor stopped');
```

Comment:
Technically, you should disconnect the events here too. Not that it matters because the whole engine is going down and the appdatabase won't be emitting more signals, but it's good practice and it helps with leaks.